Basic/DyanmicPrefabs/Assets/Scripts/HostJoinUI.cs

```diff
@@ -0,0 +1,147 @@
+using System;
+using System.Text.RegularExpressions;
+using Unity.Netcode;
+using Unity.Netcode.Transports.UTP;
+using UnityEngine;
+using UnityEngine.UIElements;
+
+public class HostJoinUI : MonoBehaviour
+{
+    // UI Documents
+    [SerializeField]
+    UIDocument m_IPMenuUIDocument;
+    [SerializeField]
+    UIDocument m_ConnectionTypeUIDocument;
+
+    // UI Roots
+    VisualElement m_IPMenuUIRoot;
+    VisualElement m_ConnectionTypeUIRoot;
+
+    // UI Elements
+    TextField m_IPInputField;
+    TextField m_PortInputField;
+    Button m_ButtonHost;
+    Button m_ButtonServer;
+    Button m_ButtonClient;
+    Button m_ButtonDisconnect;
+    Label m_ConnectionTypeLabel;
+
+    void Awake()
+    {
+        SetupIPInputUI();
+
+        // register buttons to methods using callbacks for when they're clicked 
+        m_ButtonHost.clickable.clicked += StartHost;
+        m_ButtonServer.clickable.clicked += StartServer;
+        m_ButtonClient.clickable.clicked += StartClient;
+        m_ButtonDisconnect.clickable.clicked += Disconnect;
+    }
+
+    void OnDestroy()
+    {
+        // un-register buttons from methods using callbacks for when they're clicked 
+        m_ButtonHost.clickable.clicked -= StartHost;
+        m_ButtonServer.clickable.clicked -= StartServer;
+        m_ButtonClient.clickable.clicked -= StartClient;
+        m_ButtonDisconnect.clickable.clicked -= Disconnect;
+
+        NetworkManager.Singleton.OnClientDisconnectCallback -= SingletonOnOnClientDisconnectCallback;
+    }
+
+    void Start()
+    {
+        SetUIElementVisibility(m_IPMenuUIRoot, true);
+        SetUIElementVisibility(m_ConnectionTypeUIRoot, false);
+
+        NetworkManager.Singleton.OnClientDisconnectCallback += SingletonOnOnClientDisconnectCallback;
+    }
+
+    void SingletonOnOnClientDisconnectCallback(ulong obj)
+    {
+        SetUIElementVisibility(m_IPMenuUIRoot, false);
+        SetUIElementVisibility(m_ConnectionTypeUIRoot, true);
```

Comment:
Why are we doing this here? Is the goal to reset the UI to allow a new connection when we get disconnected? If so, shouldn't the SetUIElementVisibility parameters be reversed? And we should also make sure to not do it on the server when a different client disconnects

Reply:
Oh yeah, these are backwards! And that's a good callout--I'm going to be doing some cleanup on this once I merge Fernando's work into mine, so I'll address this in that refactor :)

Reply:
Resolving this comment! @LPLafontaineB feel free to make another if you have more comments :)